### PR TITLE
Upgrade dependencies

### PR DESCRIPTION
**Pull request by sindresorhus**

### CODE REVIEW

File: benchmark/index.ts [hunk 1]
```diff
@@ -67,44 +67,44 @@ const suite = new Benchmark.Suite();
 // Benchmarking
 suite.add('got - promise', {
 	defer: true,
-	fn: async (deferred: {resolve(): void}) => {
+	fn: async (deferred: {resolve: () => void}) => {
```
Comment (sindresorhus) [inline benchmark/index.ts:70]: TS team recommends this style as it's slightly stricter.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/method-signature-style.md

File: package.json [hunk 2]
```diff
@@ -116,13 +116,11 @@
 		],
 		"rules": {
 			"@typescript-eslint/no-empty-function": "off",
-			"@typescript-eslint/no-base-to-string": "off",
 			"node/prefer-global/url": "off",
 			"node/prefer-global/url-search-params": "off",
-			"@typescript-eslint/prefer-readonly-parameter-types": "off",
+			"import/no-anonymous-default-export": "off",
 			"@typescript-eslint/no-floating-promises": "off",
```
Comment (sindresorhus) [inline package.json:122]: @szmarczak We need to enable this rule. It will help us catch unhandled exceptions errors. I think it's better if you do it though, as you can better answer whether something is safe to leave as floating or not. See https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md

File: source/core/index.ts [hunk 3]
```diff
@@ -836,6 +832,8 @@ export default class Request extends Duplex implements RequestEvents<Request> {
 
 				options.cookieJar = {
 					setCookie,
+					// TODO: Fix this when upgrading to TypeScript 4.
+					// @ts-expect-error
```
Comment (sindresorhus) [inline source/core/index.ts:836]: This is probably solvable, but I'm not sure and I need sleep.

File: source/utils/deep-freeze.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import is from '@sindresorhus/is';
 
-export default function deepFreeze<T extends object>(object: T): Readonly<T> {
+export default function deepFreeze<T extends Record<string, any>>(object: T): Readonly<T> {
```
Comment (sindresorhus) [inline source/utils/deep-freeze.ts:3]: This is also slightly stricter than `object`, which is why we should prefer it.

File: source/core/index.ts [hunk 5]
```diff
@@ -60,15 +60,13 @@ export interface Agents {
 export const withoutBody: ReadonlySet<string> = new Set(['GET', 'HEAD']);
 
 export interface ToughCookieJar {
-	getCookieString(currentUrl: string, options: {[key: string]: unknown}, cb: (err: Error | null, cookies: string) => void): void;
-	getCookieString(url: string, callback: (error: Error | null, cookieHeader: string) => void): void;
-	setCookie(cookieOrString: unknown, currentUrl: string, options: {[key: string]: unknown}, cb: (err: Error | null, cookie: unknown) => void): void;
-	setCookie(rawCookie: string, url: string, callback: (error: Error | null, result: unknown) => void): void;
+	getCookieString: ((currentUrl: string, options: {[key: string]: unknown}, cb: (err: Error | null, cookies: string) => void) => void) & ((url: string, callback: (error: Error | null, cookieHeader: string) => void) => void);
```
Comment (szmarczak) [inline source/core/index.ts:63]: Why this change?

File: source/core/index.ts [hunk 6]
```diff
@@ -279,10 +277,7 @@ export interface Response<T = unknown> extends PlainResponse {
 }
 
 export interface RequestEvents<T> {
-	on(name: 'request', listener: (request: http.ClientRequest) => void): T;
-	on<R extends Response>(name: 'response', listener: (response: R) => void): T;
-	on<R extends Response, N extends NormalizedOptions>(name: 'redirect', listener: (response: R, nextOptions: N) => void): T;
-	on(name: 'uploadProgress' | 'downloadProgress', listener: (progress: Progress) => void): T;
+	on: ((name: 'request', listener: (request: http.ClientRequest) => void) => T) & (<R extends Response>(name: 'response', listener: (response: R) => void) => T) & (<R extends Response, N extends NormalizedOptions>(name: 'redirect', listener: (response: R, nextOptions: N) => void) => T) & ((name: 'uploadProgress' | 'downloadProgress', listener: (progress: Progress) => void) => T);
```
Comment (szmarczak) [inline source/core/index.ts:280]: This looks ugly.

File: source/core/index.ts (hunk 5 above)
Comment (sindresorhus) [reply source/core/index.ts:63]: It's the linter => https://github.com/sindresorhus/got/pull/1345#discussion_r449921708

File: source/utils/deep-freeze.ts (hunk 4 above)
Comment (szmarczak) [reply source/utils/deep-freeze.ts:3]:
```suggestion
export default function deepFreeze<T extends Record<string, unknown>>(object: T): Readonly<T> {
```

?

File: source/core/index.ts (hunk 6 above)
Comment (sindresorhus) [reply source/core/index.ts:280]: I think we can move the `&` to a separate line.

File: source/utils/deep-freeze.ts (hunk 4 above)
Comment (sindresorhus) [reply source/utils/deep-freeze.ts:3]: It has to be `any` because it's spreading. It's a dumb TS thing. 

https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/#higher-order-type-inference-from-generic-constructors See the `{ [k: string]: unknown } is no longer a wildcard assignment target` heading.

File: source/core/index.ts (hunk 6 above)
Comment (sindresorhus) [reply source/core/index.ts]: Done: https://github.com/sindresorhus/got/pull/1345/commits/4dfe3e808558251ada2b0be10b2767c2098ad310

File: source/utils/deep-freeze.ts (hunk 4 above)
Comment (szmarczak) [reply source/utils/deep-freeze.ts:3]: I don't know what this exactly means, but this works:

```ts
import is from '@sindresorhus/is';

const isArray = is.array as (arg: unknown) => arg is unknown[];

export default function deepFreeze<T extends Record<string, unknown> | unknown[]>(object: T): Readonly<T> {
	for (const value of Object.values(object)) {
		if (is.plainObject(value) || isArray(value)) {
			deepFreeze(value);
		}
	}

	return Object.freeze(object);
}
```

and this does too:

```ts
import is from '@sindresorhus/is';

export default function deepFreeze<T extends Record<string, unknown>>(object: T): Readonly<T> {
	for (const value of Object.values(object)) {
		if (is.plainObject(value) || is.array(value)) {
			deepFreeze(value as Record<string, unknown>);
		}
	}

	return Object.freeze(object);
}
```

File: package.json (hunk 2 above)
Comment (szmarczak) [reply package.json]: Well I've fixed it but it says that [this](https://github.com/sindresorhus/got/blob/b51d836f27b3dec16871ec3367d363607af42bb2/test/progress.ts#L15-L42) returns a Promise even though we clearly see that it doesn't.

I think we should revert this PR to 6b7e0ec and go with this option off. Maybe make an issue about this.

Comment (szmarczak) [reply package.json]: It's really a pity you cannot mark a promise as `safe` (non-throwing). If it was, it should skip the error if there's a `return;` after the promise since it's an async task. Like some sort of events.

File: source/utils/deep-freeze.ts (hunk 4 above)
Comment (sindresorhus) [reply source/utils/deep-freeze.ts:3]: 1. Ah right. I missed that it also accepts arrays. Still fails though:

```
source/create.ts:292:59 - error TS2345: Argument of type 'InstanceDefaults' is not assignable to parameter of type 'unknown[] | Record<string, unknown>'.
  Type 'InstanceDefaults' is not assignable to type 'Record<string, unknown>'.
    Index signature is missing in type 'InstanceDefaults'.

292  	value: defaults.mutableDefaults ? defaults : deepFreeze(defaults),
```

2. Obviously that passes as you're using `as` which force-coerces it to whatever you want. That's not a proper solution though.

File: package.json (hunk 2 above)
Comment (sindresorhus) [reply package.json]: > It's really a pity you cannot mark a promise as safe (non-throwing).

Such an annotation would be useless as anything can throw in JS. Even accessing a defined property can throw (proxies/getter).

The good thing about this `void` annotation is that it forces you to think about the problem and decide whether it's reasonably safe to keep the promise floating or not. It's usually not, per my comment that anything can potentially throw.

Comment (sindresorhus) [reply package.json]: > Well I've fixed it but it says that this returns a Promise even though we clearly see that it doesn't.

Just add an ESLint ignore comment on it. Something there are false-positives if a type is some kind of `any` as it could potentially be a Promise then, TS doesn't know.

Comment (sindresorhus) [reply package.json]: Travis is still failing.

File: source/core/index.ts [hunk 7]
```diff
@@ -862,7 +862,7 @@ export default class Request extends Duplex implements RequestEvents<Request> {
 									result.catch(handler);
 								} else if (event === 'abort') {
 									// eslint-disable-next-line promise/prefer-await-to-then
-									result.then((request: unknown): void => {
+									void result.then((request: unknown): void => {
```
Comment (sindresorhus) [inline source/core/index.ts:865]: Are you 100% sure `result` will never reject at this point? Maybe do an async-IIFE with a try/catch just to be sure?

File: source/core/index.ts [hunk 8]
```diff
@@ -1302,7 +1305,7 @@ export default class Request extends Duplex implements RequestEvents<Request> {
 		const responseEventName = options.cache ? 'cacheableResponse' : 'response';
 
 		request.once(responseEventName, (response: IncomingMessageWithTimings) => {
-			this._onResponse(response);
+			void this._onResponse(response);
```
Comment (sindresorhus) [inline source/core/index.ts:1308]: Are you sure `_onResponse` could never reject? The code in it is quite complicated so I don't think we should assume so. Better to be safe.

File: source/utils/deep-freeze.ts (hunk 4 above)
Comment (szmarczak) [reply source/utils/deep-freeze.ts:3]: Yeah, you're right. Let's go with any.

File: source/core/index.ts (hunk 8 above)
Comment (szmarczak) [reply source/core/index.ts:1308]: The only possibility is when the `response` object is not an `IncomingMessage`. In that case it will be unhandled rejection. The solution would be to

```js
try {
	await this._onResponse(response);
} catch (error) {
	this._beforeError(error);
}
```

or just add the `try/catch` global clause inside `_onResponse`.

So, what should I do?

File: source/core/index.ts (hunk 7 above)
Comment (szmarczak) [reply source/core/index.ts:865]: If `result` throws then the error is caught here:

https://github.com/sindresorhus/got/blob/545e6a239ffcc4f10599565c2bfbd512d3bbcd43/source/core/index.ts#L1516-L1517

Since in that case `fn` would be `this._createCacheableRequest` which returns `result` (the `request` event).

But JS may not see that it's caught **later** and may throw unhandled rejection. The fix is just to add `.catch(() => {})`.

File: package.json (hunk 2 above)
Comment (szmarczak) [reply package.json]: Actually it's the `checkEvents` function was improperly typed.

File: source/core/index.ts (hunk 8 above)
Comment (sindresorhus) [reply source/core/index.ts:1310]: Wrapping the whole function body would make the code hard to read, instead I would make a wrapper function that try/catches `_beforeError` and use that one instead everywhere.

- `_beforeErrorBase` - no try/catch
- `_beforeError` - try catches `_beforeErrorBase` and is the one we actually use.

File: source/core/index.ts [hunk 9]
```diff
@@ -861,10 +861,12 @@ export default class Request extends Duplex implements RequestEvents<Request> {
...
 									result.catch(handler);
 								} else if (event === 'abort') {
+									// The empty catch is needed here in case when
+									// it rejects before it's `await`ed in `_makeRequest`.
 									// eslint-disable-next-line promise/prefer-await-to-then
-									void result.then((request: unknown): void => {
+									result.then((request: unknown): void => {
 										(request as ClientRequest).once('abort', handler);
-									});
+									}).catch(() => {});
```
Comment (sindresorhus) [inline source/core/index.ts]:
```js
(async () => {
	try {
		let request = await result;
		(request as ClientRequest).once('abort', handler);
	} catch {}
})();
```

Comment (sindresorhus) [reply source/core/index.ts]: I haven't tried, but if this works, use this:

```js
(async () => {
	try {
		let request: ClientRequest = await result;
		request.once('abort', handler);
	} catch {}
})();
```

Comment (szmarczak) [reply source/core/index.ts]: Should definitely work. Will fix.

File: source/core/index.ts (hunk 8 above)
Comment (szmarczak) [reply source/core/index.ts:1310]: `_beforeError` never throws :P
